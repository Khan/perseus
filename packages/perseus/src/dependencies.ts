import * as React from "react";

import type {PerseusDependencies, PerseusDependenciesV2} from "./types";

let _dependencies: PerseusDependencies | null | undefined = null;

export const setDependencies = (dependencies: PerseusDependencies) => {
    _dependencies = dependencies;
};

export const getDependencies = (): PerseusDependencies => {
    if (_dependencies) {
        return _dependencies;
    }

    throw new Error(
        [
            "Perseus has not been provided required dependencies.",
            "setDependencies(dependencies) must be called first.",
            "Make sure Perseus is being imported from javascript/perseus/perseus.js.",
        ].join("\n"),
    );
};

export type DependencyProps = Partial<
    React.ContextType<typeof DependenciesContext>
>;

export const DependenciesContext = React.createContext<PerseusDependenciesV2>({
    analytics: {onAnalyticsEvent: async () => {}},
});

/**
 * useDependencies provides easy access to the current Perseus dependencies.
 */
export const useDependencies = (): PerseusDependenciesV2 => {
    const deps = React.useContext(DependenciesContext);

    return deps;
};
