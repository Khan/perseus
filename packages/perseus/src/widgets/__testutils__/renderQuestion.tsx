import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
import {render} from "@testing-library/react";
import * as React from "react";

import {
    testDependenciesV2,
    testDependencies,
    // eslint-disable-next-line import/no-relative-packages
} from "../../../../../testing/test-dependencies";
import {
    DependenciesContext,
    useDependencies,
    setDependencies,
} from "../../dependencies";
import * as Perseus from "../../index";
import {mockStrings} from "../../strings";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing";

import type {APIOptions} from "../../types";
import type {PerseusRenderer} from "@khanacademy/perseus-core";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type RenderResult = ReturnType<typeof render>;

type ExtraProps = Omit<PropsFor<typeof Perseus.Renderer>, "strings">;

export const renderQuestion = (
    question: PerseusRenderer,
    apiOptions: APIOptions = Object.freeze({}),
    extraProps?: ExtraProps,
): {
    container: HTMLElement;
    renderer: Perseus.Renderer;
    rerender: (question: PerseusRenderer, extraProps?: ExtraProps) => void;
    unmount: RenderResult["unmount"];
} => {
    setDependencies(testDependencies);
    registerAllWidgetsForTesting();

    let renderer: Perseus.Renderer | null = null;
    const {container, rerender, unmount} = render(
        <RenderStateRoot>
            <DependenciesContext.Provider value={testDependenciesV2}>
                <Renderer
                    ref={(node) => (renderer = node)}
                    question={question as any}
                    apiOptions={apiOptions}
                    extraProps={{
                        ...extraProps,
                        strings: mockStrings,
                    }}
                />
            </DependenciesContext.Provider>
        </RenderStateRoot>,
    );
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
    if (!renderer) {
        throw new Error(`Failed to render!`);
    }
    const renderAgain = (
        question: PerseusRenderer,
        extraProps?: ExtraProps,
    ) => {
        rerender(
            <RenderStateRoot>
                <DependenciesContext.Provider value={testDependenciesV2}>
                    <Renderer
                        ref={(node) => (renderer = node)}
                        question={question}
                        apiOptions={apiOptions}
                        extraProps={{
                            ...extraProps,
                            strings: mockStrings,
                        }}
                    />
                </DependenciesContext.Provider>
            </RenderStateRoot>,
        );
        if (!renderer) {
            throw new Error(`Failed to rerender!`);
        }
    };

    return {container, renderer, rerender: renderAgain, unmount};
};

const Renderer = React.forwardRef<
    Perseus.Renderer,
    {
        question: PerseusRenderer;
        apiOptions: APIOptions;
        extraProps?: PropsFor<typeof Perseus.Renderer>;
    }
>((props, ref) => {
    const dependencies = useDependencies();
    return (
        <Perseus.Renderer
            ref={ref}
            content={props.question.content}
            images={props.question.images}
            widgets={props.question.widgets}
            problemNum={0}
            apiOptions={props.apiOptions}
            strings={mockStrings}
            {...props.extraProps}
            {...dependencies}
        />
    );
});
