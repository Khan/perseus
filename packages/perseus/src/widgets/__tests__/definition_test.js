// @flow
import {screen} from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import "@testing-library/jest-dom"; // Imports custom matchers

import {testDependencies} from "../../../../../testing/test-dependencies.js";
import * as Dependencies from "../../dependencies.js";

import {renderQuestion} from "./renderQuestion.jsx";

const question = {
    content:
        "Read the excerpt and answer the question below. \n\nThe Governor and Council of the Massachusetts had much conference many days; and at last . . . . concluded a peace and friendship with [[\u2603 definition 1]], upon these conditions.",
    images: {},
    widgets: {
        "definition 1": {
            graded: true,
            version: {
                major: 0,
                minor: 0,
            },
            static: false,
            type: "definition",
            options: {
                definition: "Definition text",
                togglePrompt: "the Pequots",
                static: false,
            },
            alignment: "default",
        },
    },
};

describe("Definition widget", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("should have a default snapshot", () => {
        // Arrange & Act
        const {container} = renderQuestion(question);

        // Assert
        expect(container).toMatchSnapshot("first render");
    });

    it("should have an open state snapshot", () => {
        // Arrange
        const {container} = renderQuestion(question);

        // Act
        const definitionAnchor = screen.getByText("the Pequots");
        userEvent.hover(definitionAnchor);
        jest.advanceTimersByTime(250);

        // Assert
        expect(container).toMatchSnapshot("open state");
    });

    it("should display the definition on hover", () => {
        // Arrange
        renderQuestion(question);

        // Act
        const definitionAnchor = screen.getByText("the Pequots");
        userEvent.hover(definitionAnchor);
        jest.advanceTimersByTime(250);

        // Assert
        const tooltip = screen.getByRole("tooltip");
        expect(tooltip).toBeVisible();
        expect(tooltip).toHaveTextContent("Definition text");
    });

    it("should display the definition on click", () => {
        // Arrange
        renderQuestion(question);

        // Act
        const definitionAnchor = screen.getByText("the Pequots");
        userEvent.click(definitionAnchor);
        const tooltip = screen.getByRole("tooltip");

        // Assert
        expect(tooltip).toBeVisible();
        expect(tooltip).toHaveTextContent("Definition text");
    });

    it("should show via focus by the tab key", () => {
        // Arrange
        renderQuestion(question);

        // Act - Tab in to set focus
        userEvent.tab();
        jest.advanceTimersByTime(250);

        // Assert
        const tooltip = screen.getByRole("tooltip");
        expect(tooltip).toBeVisible();
        expect(tooltip).toHaveTextContent("Definition text");
    });

    it("should hide and blur by the tab key", () => {
        // Arrange
        renderQuestion(question);

        // Act - Tab in to set focus, tab out to blur
        userEvent.tab();
        jest.advanceTimersByTime(250);
        userEvent.tab();
        jest.advanceTimersByTime(250);

        // Assert
        expect(screen.queryByRole("tooltip")).toBeNull();
    });

    it("should dimiss by a click when showing", () => {
        renderQuestion(question);

        // Act
        // Click on the anchor
        const definitionAnchor = screen.getByText("the Pequots");
        userEvent.click(definitionAnchor);

        // Move the mouse away and make sure text is still visible
        userEvent.unhover(definitionAnchor);
        jest.advanceTimersByTime(250);

        // Click to elsewhere, tooltip is hidden
        userEvent.click((document.body: any));
        jest.advanceTimersByTime(250);

        // Assert
        expect(screen.queryByRole("tooltip")).toBeNull();
    });

    it("should not affect answerable", () => {
        // Arrange and act
        const {renderer} = renderQuestion(question);
        const result = renderer.scoreWidgets();

        // Assert
        expect(result["definition 1"]).toMatchObject({
            type: "points",
            earned: 0,
            total: 0,
            message: null,
        });
    });
});
