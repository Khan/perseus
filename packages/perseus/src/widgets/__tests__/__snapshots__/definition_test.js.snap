// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Definition widget should have a default snapshot: first render 1`] = `
<div>
  <span
    class="mock-KatexProvider"
  >
    <div
      class="perseus-renderer perseus-renderer-responsive"
    >
      <div
        class="paragraph"
        data-perseus-paragraph-index="0"
      >
        <div
          class="paragraph"
        >
          Read the excerpt and answer the question below. 
        </div>
      </div>
      <div
        class="paragraph"
        data-perseus-paragraph-index="1"
      >
        <div
          class="paragraph"
        >
          The Governor and Council of the Massachusetts had much conference many days; and at last . . . . concluded a peace and friendship with 
          <div
            class="perseus-widget-container widget-nohighlight widget-inline"
          >
            <span
              aria-describedby="uid-tooltip-0-aria-content"
              class="perseus-widget-definition"
              tabindex="0"
            >
              <div
                class="default_xu2jcg-o_O-inlineStyles_txdle"
              >
                the Pequots
              </div>
            </span>
          </div>
          , upon these conditions.
        </div>
      </div>
    </div>
  </span>
</div>
`;

exports[`Definition widget should have an open state snapshot: open state 1`] = `
<div>
  <span
    class="mock-KatexProvider"
  >
    <div
      class="perseus-renderer perseus-renderer-responsive"
    >
      <div
        class="paragraph"
        data-perseus-paragraph-index="0"
      >
        <div
          class="paragraph"
        >
          Read the excerpt and answer the question below. 
        </div>
      </div>
      <div
        class="paragraph"
        data-perseus-paragraph-index="1"
      >
        <div
          class="paragraph"
        >
          The Governor and Council of the Massachusetts had much conference many days; and at last . . . . concluded a peace and friendship with 
          <div
            class="perseus-widget-container widget-nohighlight widget-inline"
          >
            <span
              aria-describedby="uid-tooltip-1-aria-content"
              class="perseus-widget-definition"
              tabindex="0"
            >
              <div
                class="default_xu2jcg-o_O-inlineStyles_txdle"
              >
                the Pequots
              </div>
            </span>
          </div>
          , upon these conditions.
        </div>
      </div>
    </div>
  </span>
</div>
`;
